### ISSUE TITLE
Suggestion: profiling

### ISSUE THREAD

**Issue by philipturner** [artyom-beilis/dlprimitives#12]
Apple has highly optimized DL primitives in Metal performance shaders. By partially de-compiling Metal Performance Shaders, I saw an insane amount of permutations. They optimized for all sorts of edge cases. I saw the term "winograd" once or twice in function names. I tried comparing custom Metal shaders to Apple's MPS and mine were terrible, but I imagine you have more time to thoroughly investigate performance deltas.

Given that Metal works on most AMD and Intel GPUs, it would be wise to run your OpenCL code on macOS and compare your performance to Apple's. That would ensure your kernels utilize the GPU as much as physically possible. Another suggestion is to try comparing DirectML, although I suspect that Apple is more optimized due to the sheer number of permutations they created. You can examine the DirectML source code to see if Microsoft takes the permutation approach too.

**Comment by artyom-beilis** [comment]
In general, I use OpenCL to make things platform independent so Metal and DirectML aren't relevant.

I hadn't tested dlprimitives on Mac OS X since I don't have one - but you are welcome to compare. Regarding DirectML I just don't have capacity to check against it.

I have good baseline reference of cuda+cudnn for nVidia and I'll have AMD's MIOpen on GCN and RNDA (once they release it)

There is also oneDNN for Intel but it has poor performance if channel first layout: https://github.com/oneapi-src/oneDNN/issues/1194 

Bottom line I know there are lots of improvements in performance can be done. I still don't reach cuDNN level of performance but at this point it is mostly good enough.

I'd be glad to receive help in writing better kernels, adopting code from other projects.

At this point I try to cover much more operators, make pytorch/dlprimitives much more usable and give decent inference support.

**Comment by philipturner** [comment]
> At this point I try to cover much more operators, make pytorch/dlprimitives much more usable and give decent \*\*inference\*\* support.

Training is the most meaningful use case for this stuff. Are you planning to restrict implementations to inference?

**Comment by artyom-beilis** [comment]
> Training is the most meaningful use case for this stuff

It is _one of_ important use cases and I work on it (see pytorch backend I work on)

>  Are you planning to restrict implementations to inference?

By no means!

But I want to have lightweight inference tool: see https://github.com/artyom-beilis/dlprimitives/tree/onednn_integration
I work on loading ONNX model directly by dlprimitives that would make it much more easy to deploy inference in cross platform way with minimal dependencies.

**Comment by artyom-beilis** [comment]
I meant https://github.com/artyom-beilis/dlprimitives/tree/onnx_api

Onnx support. oneDNN is useless for channel first layout

**Comment by philipturner** [comment]
Do you have a short benchmark suite? If so, I could translate it to Metal Performance Shaders and run it on a few GPU models. Can you test your code on an Intel UHD 630 or Iris Plus? Or what AMD GPUs do you have? The UHD 630 is the one I have instant access to; the others are ones my friends have.

**Comment by artyom-beilis** [comment]
> Do you have a short benchmark suite?
> If so, I could translate it to Metal Performance Shaders and run it on a few GPU models. 

Yes: https://github.com/artyom-beilis/dlprimitives/blob/master/docs/build.md#benchmarking
Also there is `dlprim_flops` for specific kernel performance measurements

> Can you test your code on an Intel UHD 630 or Iris Plus?
> Or what AMD GPUs do you have? The UHD 630 is the one I have instant access to; the others are ones my friends have.

see: https://github.com/artyom-beilis/dlprimitives#tested-gpus

On Intel side I test it on UHD 530 that is basically same as 630 up to slight clock changes.

To be hones Intel GPUs with 400GFlops quite useless for trainin. they have ~ same performance as CPU training and actually CPU training with mklDNN may be way faster - this what frameworks like pytorch/tf use. Also with modern multi-core CPU the built-in GPU would likely be slower. Maybe when I implement float16 support with ~700GFlops it would be little bit more useful for inference.

**Comment by philipturner** [comment]
I agree the UHD 630 is kind of useless. I could also run your benchmark suite on it on Windows to provide the most accurate comparison. I think there’s a difference between HD 530 and UHD 530, so be careful with naming if that’s not the case.

**Comment by philipturner** [comment]
I was expecting to have pre-recorded benchmarks of how many milliseconds each test takes on a particular GPU. Then, I would make a simple command-like utility to compare performance on macOS. I was not expecting to have to set up all the OpenCL stuff by myself and modify the code to collect benchmarks. It’s not worth my time now if I have to do all of that.

**Comment by artyom-beilis** [comment]
I suggest reading docs and some blog posts 

There are plenty of benchmark results. 

However since every gpu and system little bit different it is better to run comparison on same machine

**Comment by philipturner** [comment]
I don't understand why there's an order of magnitude performance difference between Caffe and PlaidML. The difference is the same across weaker and stronger GPUs, so it seems like Caffe is using the GPU. I'm confused because I thought PyTorch trained exclusively on the CPU except on Nvidia. Also, what do the percentages mean? Do larger percents mean more efficiency or more execution time, and is that convention consistent across all of your benchmarks?

**Comment by artyom-beilis** [comment]
> I don't understand why there's an order of magnitude performance difference between Caffe and PlaidML

Because PlaidML has very poor performance

> , so it seems like Caffe is using the GPU

Caffe has OpenCL, cuda and hip/rocm versions. But caffe isn't developed any more and its memory use is not optimal at all

> I'm confused because I thought PyTorch trained exclusively on the CPU except on Nvidia.

There is:

- Hip/ROCm support of AMD GPUs of older generation (GCN), no RDNA support yet and only Linux supoorted
- pytorch-dlprim :-)
